Add Judger implementation

candicez03 opened the pull request:
Adds the judge package for submission testing and grading.

To judge a submission, create an instance of `Judger` and call `.judge(Submission)`. The `Judger` then proceeds to:
1. Get a `SourceChecker` to check if the source code contains predefined illegal segments, depending on the language of the submission. If the source code does contain any of the segments, or the language is not supported, the Judger updates the status of the submission accordingly and judging ends.
2. Get a `SourceLauncher` by calling `SourceLauncherService.getSourceLauncher(...)` and try to setup the program. The `SourceLauncher` is responsible for creating a temporary file for the program and setting up the program (e.g. compilation). If the language is not supported, or some internal exceptions occur, the Judger updates the status of the submission accordingly and judging ends.
3. Repeatedly call `Tester.test(...)` for each of the testcase in the problem, which returns the result of the testcase run. For each batch of the problem, if one of the testcase run fails, the rest are skipped. Otherwise (if all testcases pass), the submission gets the points of the batch. All programs are launched via `GlobalChildProcessService.launchChildProcess(..)`. 
    - `GlobalChildProcessService.validateActiveChildProcesses()` updates the memory usage of current child processes and terminates any child process that exceeds its memory usage limit. This method is called whenever a child process is launched via `GlobalChildProcessService.launchChildProcess(..)`. In addition, GlobalChildProcessService has an instance of `GlobalChildProcessService.ChildProcessMonitor` that reapeatedly calls this method with a predefined frequency.
4. Update the submission's stats (score, run duration, memory usage, status), and write it back to the database. Currently, the entities in `src/entities/` are a simplified version, thus no actual db writing is implemented yet.

See `judge.class.violet.html` for further details.

candicez03 committed:
Basic judging features

- Checks (methods to be implemented), sets up, executes, and grades submissions
- Todo: implementing checking methods, detecting memory/output limit exceeded and memory usage, adding support for more languages

candicez03 committed:
UML update on judge

candicez03 committed:
added OUTPUT_LIMIT_EXCEEDED detection, updated judge UML accordingly

candicez03 committed:
Merge branch 'master' into judge

candicez03 committed:
updated param types and namings

- changed memory limit in bytes to in kilobytes
- changed time/memory/output limits' data types from long to int

candicez03 committed:
Added memory tracking related features

- GlobalChildProcessService: periodically retrieves memory usage of each child process, kills the process when it exceeds its assigned memory limit
- todo: improve efficiency (its very slow now...), do more testing and fix potential bugs, complete clean checkers

candicez03 committed:
filled in clean checker methods

candicez03 committed:
updated judge uml

candicez03 committed:
Moves SourceLauncherService.java into judge/services

candicez03 committed:
updated Judger.java, recovered resources uml

Judger.judge(Submission) now returns a Submission

candicez03 committed:
Removed judge uml file from repo

beepboop271 commented on src/judge/test/Main.java:
```diff
@@ -0,0 +1,203 @@
+    long start = System.currentTimeMillis();
+    for (Submitter s : submitters) {
+      s.run();
+    }
```
Don't run your test submissions sequentially. It's slow and does not represent a more realistic load where the web server threads can call into judge concurrently. Adding this many submissions at the same time would be more like a worst-case load for our program though.

beepboop271 commented on src/judge/test/Main.java:
```diff
@@ -0,0 +1,203 @@
+      new Timestamp(System.currentTimeMillis())
+    );
+
+    File directory = new File("cache/judge/");
```
`cache` is a bad folder name for this purpose. A cache is for something you quickly read from often. You are reading from the files once and deleting them right after.

beepboop271 commented on src/judge/test/Main.java:
```diff
@@ -0,0 +1,203 @@
+
+    File directory = new File("cache/judge/");
+    Judger judger = new Judger(
+      Runtime.getRuntime().availableProcessors(),
```
Might not be the best choice of thread pool size, but that doesn't matter too much for now.

beepboop271 commented on src/judge/entities/UserException.java:
```diff
@@ -0,0 +1,22 @@
+package judge.entities;
+
+@SuppressWarnings("serial")
```
Add a serial version ID (you can just start with 0) instead of suppressing the warning.

beepboop271 commented on src/judge/entities/Judger.java:
```diff
@@ -0,0 +1,234 @@
+  private void updateDatabase(TestcaseRun testcaseRun) {
+    //TODO: write to db
+    //display(testcaseRun);
+  }
```
Your code is not supposed to know about the database at all.

beepboop271 commented on src/judge/entities/Judger.java:
```diff
@@ -0,0 +1,234 @@
+    this.tempFileDirectory = tempFileDirectory;
+  }
+
+  public Submission judge(Submission submission) {
```
This method is way too long.

beepboop271 commented on src/judge/entities/Judger.java:
```diff
@@ -0,0 +1,234 @@
+      this.updateDatabase(submission);
+      launcher.close();
+      return submission;
+    }
```
3 things:
- There really isn't much point to using a name other than `e` for caught exceptions in my opinion.
- You can simplify this weird structure into the following using `|`.
- You should probably move the copy pasted failure lines into another method.
```java
} catch (CompileErrorException e) {
  return fail(ExecutionStatus.COMPILE_ERROR, e, launcher, submission);
} catch (UserException | InternalErrorException e) {
  return fail(ExecutionStatus.INTERNAL_ERROR, e, launcher, submission);
}

// the copy pasted params that are the same each time go to the
// end of the param list for fail(interesting_stuff, boring_stuff)
// instead of fail(boring_stuff, interesting_stuff)
private Submission fail(status, exception, launcher, submission) {
  submission.setStatus(status);
  e.printStackTrace();
  launcher.close();
  return submission;
}
```
Note that if you follow the other advice in `SourceLauncherService`, the `launcher.close()` line would be removed from this particular catch block.

This comment applies to all the other `catch` blocks too, it's just that this one in particular made my want to comment.

beepboop271 commented on src/judge/entities/Judger.java:
```diff
@@ -0,0 +1,234 @@
+      Testcase[] testcases = batch.getTestcases();      
+      
+      // loop through each testcase in the batch
+      for (int j = 0; j < testcases.length; j++) {
```
The contents of this loop just scream "make a new method". An indent size of 2 spaces doesn't mean you can have 7 indents.

beepboop271 commented on src/judge/entities/Judger.java:
```diff
@@ -0,0 +1,234 @@
+            outputLimitKb
+          );
+          try {
+            run = caseRunFuture.get();
```
What happened to running batches concurrently?

beepboop271 commented on src/judge/entities/SourceLauncher.java:
```diff
@@ -0,0 +1,79 @@
+    this.deleteTempFile();
+  }
+
+  public void setup() throws InternalErrorException, UserException {
```
`throws InternalErrorException, CompileErrorException`

beepboop271 commented on src/judge/services/SourceLauncherService.java:
```diff
@@ -0,0 +1,75 @@
+        this.f.completeExceptionally(new UnknownLanguageException(this.submission.getLanguage()));
+      }
+    }
+  }
```
The whole point of making all the parts future based was so that the file writing and compiling can happen asynchronously if needed (though we need it to be synchronous because the program needs to be setup before running it). However, you haven't included any blocking tasks (except for the optional temp directory creation) in this part at all, and in `Judger` you call `launcher.setup()` which is the method that synchronously creates all the necessary files (on the thread created by the web server).

So just move setup into this runnable so that the actual blocking part is run in the thread pool. I guess keeping the file stuff in the thread pool should also ensure that there's no way you'd ever get hit with 100 requests on the web server at the same time causing 100 files to be written at the same time and it would be done slowly.

beepboop271 commented on src/judge/entities/PythonLauncher.java:
```diff
@@ -0,0 +1,42 @@
+
+  @Override
+  public Process launch() throws InternalErrorException {
+    ProcessBuilder builder = new ProcessBuilder("python", this.getSource().getAbsolutePath());
```
Reminder that the exact commands need to be customizable (e.g, the command might to be `python3` or perhaps an absolute executable path).

beepboop271 commented on src/judge/entities/JavaLauncher.java:
```diff
@@ -0,0 +1,77 @@
+    if (errorMsg.length() > 0) {
+      throw new CompileErrorException(errorMsg);
+    }
+  }
```
You forgot to put files into their own folder like I said to. The class files will overwrite each other. In addition, this is an example of one of the ways your PID checker could fail unless you check for the name of the process, since `javac` would appear as a new child process.

beepboop271 commented on src/judge/entities/JavaLauncher.java:
```diff
@@ -0,0 +1,77 @@
+
+  @Override
+  public Process launch() throws InternalErrorException {
+    ProcessBuilder builder = new ProcessBuilder("java", this.getSource().getAbsolutePath());
```
`java RandomJavaFile.java` does not work. You'll need to do `java -cp path Main` where `path` is the directory to the folder containing `Main.class`. Alternatively, you can have the command `java Main` and then change the working directory of the `ProcessBuilder` to be the directory containing the class file.

beepboop271 commented on src/judge/services/SourceLauncherService.java:
```diff
@@ -0,0 +1,75 @@
+ * @since 1.0.0
+ */
+
+public class SourceLauncherService {
```
Perhaps rename to `ProgramLauncher` to be more consistent with `ProgramChecker` -> `SourceChecker` -> `Java`/`PythonChecker`. You're also missing a private constructor here.

beepboop271 commented on src/judge/services/Tester.java:
```diff
@@ -0,0 +1,163 @@
+        internalErrorException.printStackTrace();
+      } catch (InterruptedException interruptedException) {
+        interruptedException.printStackTrace();
+      }
```
What is this exception handling...

beepboop271 commented on src/judge/services/Tester.java:
```diff
@@ -0,0 +1,163 @@
+
+  }
+
+}
```
Blank lines? Note that this happens many times, I'm just leaving a comment on one example to point it out.

beepboop271 commented on src/judge/services/Tester.java:
```diff
@@ -0,0 +1,163 @@
+    }
+
+    @Override
+    public void run() {
```
This method is also pretty long.

beepboop271 commented on src/judge/services/Tester.java:
```diff
@@ -0,0 +1,163 @@
+            runToReturn.setStatus(ExecutionStatus.OUTPUT_LIMIT_EXCEEDED);
+            break;
+          }
+        }
```
Why are you copying bytes one at a time when StringBuilder can use `System.arraycopy` when you append char arrays. Note that you will need to change your `new BufferedInputStream(program.getInputStream())` to use `new InputStreamReader(program.getInputStream())` (to read characters instead of bytes, and don't use a `BufferedReader` because you're doing buffering). This is fine because all program input/output should be done with ascii characters. You can just read in 1 or several kilobytes at a time (8k if you want to be the same as `BufferedReader`), not like the output limit will be very small. I believe you do this elsewhere too.

beepboop271 commented on src/judge/services/Tester.java:
```diff
@@ -0,0 +1,163 @@
+        }
+        // compare with expected output if submission hasn't received a status
+        if (runToReturn.getStatus() == ExecutionStatus.PENDING) { 
+          if (sb.toString().trim().equals(expectedOutput.trim())) { // trim whitespace
```
Don't forget to document somewhere that leading and trailing whitespaces are not significant, but whitespace between is.

beepboop271 commented on src/judge/services/GlobalChildProcessService.java:
```diff
@@ -0,0 +1,155 @@
+ * @since 1.0.0
+ */
+
+public class GlobalChildProcessService {
```
I think the name `ChildProcesses` is adequate (e.g. Collections has static methods for Collection, Files has static methods for File, Arrays has static methods for...).

beepboop271 commented on src/judge/services/GlobalChildProcessService.java:
```diff
@@ -0,0 +1,155 @@
+ */
+
+public class GlobalChildProcessService {
+  private static final int UPDATE_INTERVAL_MILLIS = 1000*1;
```
Does this have to be fixed? Perhaps maybe wait 1 second (or just suspend the thread entirely) when you're not running something and then a shorter time/wake up the thread when you are.

beepboop271 commented on src/judge/services/GlobalChildProcessService.java:
```diff
@@ -0,0 +1,155 @@
+    Process process = launcher.launch();
+    List<OsProcess> updatedProcesses = SYSTEM_MONITOR.processTree().find(CURRENT_PID).children();
+
+    LinkedHashSet<Integer> updatedProcessPids = new LinkedHashSet<Integer>();
```
Why is this a LinkedHashSet?

beepboop271 commented on src/judge/services/GlobalChildProcessService.java:
```diff
@@ -0,0 +1,155 @@
+        } catch (InterruptedException e) {
+          e.printStackTrace();
+          this.stop();
+        }
```
As mentioned above, you could try replacing this with something more intelligent, such as using `Object.wait()` and `Object.notify()`.

beepboop271 commented on src/judge/services/GlobalChildProcessService.java:
```diff
@@ -0,0 +1,155 @@
+        ProcessInfo childInfo = childProcessTree.processInfo();
+        double memoryUsedKb = childInfo.getTotalBytes()/1024.0;
+        childProcess.setMemoryUsedKb(
+          Math.max(memoryUsedKb, childProcess.getMemoryUsedKb())
```
Maybe change ChildProcess.setMemoryUsedKb to have some other name (updateMemoryUsageKb?) and always take max into account in that method so that an external class can't just set whatever it wants.

beepboop271 commented on src/judge/services/GlobalChildProcessService.java:
```diff
@@ -0,0 +1,155 @@
+  }
+
+  public static void validateActiveChildProcesses() {
+    OsProcess parentProcessTree = SYSTEM_MONITOR.processTree();
```
You should `.find` the current pid here so that you dont need to always walk from root later in the method.

beepboop271 commented on src/judge/services/GlobalChildProcessService.java:
```diff
@@ -0,0 +1,155 @@
+
+      } else {
+        ProcessInfo childInfo = childProcessTree.processInfo();
+        double memoryUsedKb = childInfo.getTotalBytes()/1024.0;
```
`childInfo.getResidentBytes()`
No need to penalize them for memory they aren't even really using.

beepboop271 commented on src/judge/services/GlobalChildProcessService.java:
```diff
@@ -0,0 +1,155 @@
+    for (OsProcess p : updatedProcesses) {
+      updatedProcessPids.add(p.processInfo().getPid());
+    }
+    int childProcessPid = GlobalChildProcessService.getNewPid(updatedProcessPids);
```
It might be nice to make it easier to swap between Java >= 9 and Java < 9 methods of getting PID.

beepboop271 reviewed (changes requested):
- I am assuming that the entities are fine.
- Many times the launched process throws an exception because its PID cannot be determined due to:
  - Various other random subprocesses being spawned (can be fixed by checking process name before considering it).
  - The program completes too quickly and exits before being able to poll its PID (ie your "a = 1" python test).
- You put public on pretty much every class for some reason
- Your subpackages make no sense and it looks like you arbitrarily put classes in them (why are source checkers and source launchers "entities" when they do things like a "service"?). You can have one judge package or you can split it up sanely, perhaps one package for *Checker, one for *Launcher...
- Might want to consider renaming some classes
- Your versions conflict with Joseph's versions
- Random blank lines near the end of files, missing `this.` and missing `Class.staticThing` in places
- Some sketchy exception handling with some copy paste there too
- You need to split up some big methods
(see comments for examples of some entries)

candicez03 committed:
improved packaging and class naming, made batch testing concurrent

- need to fix pid tracking and improve child process monitoring

candicez03 committed:
fixed pid tracking, improved child process monitoring

- the tracking logic is still kinda sketch, will try to fix later
- need to separate temp files into their own folders, fix java program compiling issue, and do more testing

candicez03 committed:
fixed temp file issues

- program files are separated into their own folders
- fixed javalauncher compiling issues
- todo: more testing, less hardcoded things (commands, etc.)

candicez03 committed:
Renamed ProgramLauncher and ProgramChecker

- renamed ProgramLauncher.java to SourceLauncherService.java
- renamed ProgramChecker.java to SourceCheckerService.java

candicez03 commented:
Fixed most issues addressed:
- Improved packaging and class naming
- Batch testing is now concurrent
- Fixed pid tracking issues, and used `Object.wait()` and `Object.notify()` to monitor child processes
- Program files are now separated into their own folders, and java submissions are able to compile and run
    - For java submissions, the class has to be named `Main`

Pending fixes:
- Make illegal strings customizable (by external classes, via a Settings object or a static method)
- Make launcher commands customizable

candicez03 committed:
Updated entities and Judger

Removed Problem from Submission's fields

candicez03 committed:
Updated Tester.java

quick fix on batch testing issue

candicez03 committed:
Fixed child process killing upon shutdown of ChildProcesses

- replaced javasysmon's killprocess tree with a loop of Process.destroyforcibly()

candicez03 committed:
Added javadoc for judge.checker and judge.launcher

candicez03 committed:
Added javadoc for custom exceptions, Tester.java, and Judger.java

- the judge package (excluding the test folder) is now all documented, except child process files
- todo: try Oshi for child process monitoring, since JavaSysMon.processTree()'s performance on windows is not so ideal

candicez03 committed:
cherry pick judge changes from submission branch

candicez03 committed:
replace javasysmon with oshi

- also added some missing ChildProcesses. in front of static fields and methods
- uncommented out test.Main submissions and also properly shut down threads in the test program

candicez03 committed:
Add javadoc for child process files

- added javadoc for ChildProcess.java and ChildProcesses.java

candicez03 committed:
Fix WA on testcases with no input

now launchChildProcess returns a ChildProcess even if the pid cannot be tracked

candicez03 committed:
Update Tester.java

Prevent the TestcaseRun returned by Tester.test() (and the SubmissionResult returned by Judger.judge()) from having a PENDING status (marked as INTERNAL_ERROR instead).

candicez03 committed:
Improve Javadoc Formatting, Updated judge/test

candicez03 committed:
Update judge/test

candicez03 committed:
Fix input writing issue
